waiter/integration/waiter/authentication_test.clj

```diff
@@ -126,3 +128,152 @@
           (finally
             (delete-token-and-assert waiter-url token)))))))
 
+(defn- retrieve-access-token
+  [realm]
+  (if-let [access-token-url-env (System/getenv "WAITER_TEST_JWT_ACCESS_TOKEN_URL")]
+    (let [access-token-url (string/replace access-token-url-env "{HOST}" realm)
+          access-token-uri (URI. access-token-url)
+          protocol (.getScheme access-token-uri)
+          authority (.getAuthority access-token-uri)
+          path (str (.getPath access-token-uri) "?" (.getQuery access-token-uri))
+          access-token-response (make-request authority path :headers {"x-iam" "waiter"} :protocol protocol)
+          _ (assert-response-status access-token-response 200)
+          access-token-response-json (-> access-token-response :body str json/read-str)]
+      (get access-token-response-json "access_token"))
+    (throw (ex-info "WAITER_TEST_JWT_ACCESS_TOKEN_URL environment variable has not been provided" {}))))
+
+(defmacro assert-auth-cookie
+  "Helper macro to assert the value of the set-cookie header."
+  [set-cookie assertion-message]
+  `(let [set-cookie# ~set-cookie
+         assertion-message# ~assertion-message]
+     (is (string/includes? set-cookie# "x-waiter-auth=") assertion-message#)
+     (is (string/includes? set-cookie# "Max-Age=") assertion-message#)
+     (is (string/includes? set-cookie# "Path=/") assertion-message#)
+     (is (string/includes? set-cookie# "HttpOnly=true") assertion-message#)))
+
+(deftest ^:parallel ^:integration-fast test-jwt-authentication-waiter-realm
+  (testing-using-waiter-url
+    (if (jwt-auth-enabled? waiter-url)
+      (let [waiter-host (-> waiter-url sanitize-waiter-url utils/authority->host)
+            access-token (retrieve-access-token waiter-host)
+            request-headers {"authorization" (str "Bearer " access-token)
+                             "host" waiter-host
+                             "x-forwarded-proto" "https"}
+            {:keys [port]} (waiter-settings waiter-url)
+            target-url (str waiter-host ":" port)
+            {:keys [body headers] :as response}
+            (make-request target-url "/waiter-auth" :disable-auth true :headers request-headers :method :get)
+            set-cookie (str (get headers "set-cookie"))
+            assertion-message (str {:headers headers
+                                    :set-cookie set-cookie
+                                    :target-url target-url})]
+        (assert-response-status response 200)
+        (is (= (retrieve-username) (str body)))
+        (is (= "jwt" (get headers "x-waiter-auth-method")) assertion-message)
+        (is (= (retrieve-username) (get headers "x-waiter-auth-user")) assertion-message)
+        (assert-auth-cookie set-cookie assertion-message))
+      (log/info "JWT authentication is disabled"))))
+
+(deftest ^:parallel ^:integration-fast test-bypass-jwt-authentication-waiter-realm
```

Comment:
Should this be named something like `test-fallback-to-alternate-auth-on-invalid-jwt-token`?  Are we testing that authentication falls back to another auth method if we provide an invalid JWT?  Is that what we want?

Reply:
JWT *could* stand on its own, meaning it can work without any other authentication method enabled.  You haven't implemented it to work that way.  That's fine.  I'm confused the way we've set up the challenges however.  See my other comment regarding challenges.